### PR TITLE
Introduce getTimeOffset API to date-range picker

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

The new getTimeOffset API allowing time offsets to be provided per date to account for daylight savings if different between start and end dates of the selected range.

Usage examples:

```
// Same as default
<DateRangePicker ... getTimeOffset={date => -1 * new Date(date).getTimezoneOffset()} />

// Parse in a custom timezone
import moment from 'moment-timezone'
<DateRangePicker ... getTimeOffset={date => -1 * moment.tz(date, 'America/Toronto')} />
```

### How has this been tested?

Added unit tests.

### Documentation changes

[*Do the changes include any API documentation changes?*]
- [x] _Yes, this change contains documentation changes._
- [ ] _No._

### Related Links

* Ticket AWSUI-19513

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- [ ] _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](CONTRIBUTING.md#public-apis)._
- [ ] _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](CONTRIBUTING.md#browsers-support)._
- [ ] _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- [ ] _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- [ ] _Changes are covered with new/existing unit tests?_
- [ ] _Changes are covered with new/existing integration tests?_
</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/date-range-picker/index.tsx [hunk 1]
```diff
@@ -109,7 +110,8 @@ const DateRangePicker = React.forwardRef(
     const { __internalRootRef } = useBaseComponent('DateRangePicker');
     checkControlled('DateRangePicker', 'value', value, 'onChange', onChange);
 
-    value = isDateOnly(value) ? value : shiftTimeOffset(value, timeOffset);
+    const normalizedTimeOffset = normalizeTimeOffset(value, getTimeOffset, timeOffset);
+    value = isDateOnly(value) ? value : shiftTimeOffset(value, normalizeTimeOffset(value, getTimeOffset, timeOffset));
```
Comment (gethinwebster) [inline src/date-range-picker/index.tsx:114]: [minor] seems to be some duplication in these two lines

**Review by gethinwebster — APPROVED** [review]

File: src/date-range-picker/index.tsx (hunk 1 above)
Comment (pan-kot) [reply src/date-range-picker/index.tsx:114]: oh, indeed 👍

File: src/__tests__/__snapshots__/documenter.test.ts.snap [hunk 2]
```diff
@@ -4809,6 +4809,17 @@ in a slight, visible lag when scrolling complex pages.",
       "optional": true,
       "type": "boolean",
     },
+    Object {
+      "description": "A function from ISO8601 date string to return the desired time offset from UTC in minutes.
```
Comment (gethinwebster) [inline src/__tests__/__snapshots__/documenter.test.ts.snap:4813]: "A function _that receives_ an ISO8601 date string _and returns_ the desired..."

File: src/__tests__/__snapshots__/documenter.test.ts.snap [hunk 3]
```diff
@@ -5103,7 +5114,8 @@ display and produce values.
 Has no effect when \`dateOnly\` is true.
 
 Default: the user's current time offset as provided by the browser.
-",
+
+The property is deprecated, use getTimeOffset instead.",
```
Comment (stirache) [inline src/__tests__/__snapshots__/documenter.test.ts.snap:5118]:
```suggestion
This property is deprecated. Use getTimeOffset instead.
```

Comment (pan-kot) [reply src/__tests__/__snapshots__/documenter.test.ts.snap:5118]: Oh, thanks. I think it was applied to the documenter output and then overwritten. I will update.

File: src/date-range-picker/interfaces.ts [hunk 4]
```diff
@@ -89,8 +89,21 @@ export interface DateRangePickerBaseProps {
...
+
+  /**
+   * A function that receives an ISO8601 date string and returns the desired time offset from UTC in minutes.
+   * Use it to define time relative to the desired timezone.
+   *
+   * Has no effect when `dateOnly` is true.
+   *
+   * Default: the user's current time offset as provided by the browser.
+   */
+  getTimeOffset?: (date: string) => number;
```
Comment (FlorianDr) [inline src/date-range-picker/interfaces.ts:106]:
```suggestion
  getTimeOffset?: (date: Date) => number;
```

**Review by gethinwebster — APPROVED** [review]

**Pull request merged**